Add initial scaffold for content development (#5)
software-gardening/almanac: CONTRIBUTING.md

# Contributing

First of all, thank you for contributing to the Software Gardening Almanac! 🎉 💯 🪴
We're absolutely grateful for the kindness of your effort applied to grow this project into the best that it can be.

This document contains guidelines on how to most effectively contribute to this project.

> If you are stuck or would like clarifications, please feel free to ask any questions or for help.

Reviewer: Do you have any specific mechanism in mind to ask these questions? GitHub issues, discussions, something else?

Author: Thanks, good point! I wasn't sure about this and hesitated about being too prescriptive about how one should ask for help. In my opinion, anyone offering contributions should feel welcome to ask for help where they are (within an issue, a PR, discussion, chat, etc). I'll spell that out more clearly with an addition.

Author: I added some content here to help with this. Resolving for now.

This project is governed by our [Code of Conduct (CoC) policy](https://github.com/software-gardening/almanac?tab=coc-ov-file) (located [here](https://github.com/software-gardening/.github/blob/main/CODE_OF_CONDUCT.md)).
By participating, you are expected to uphold this code.
Please report unacceptable behavior by following the procedures listed under the [CoC Enforcement section](https://github.com/software-gardening/almanac?tab=coc-ov-file#enforcement).

## Security

Please see our [Security policy](https://github.com/software-gardening/almanac?tab=security-ov-file) (located [here](https://github.com/software-gardening/.github/blob/main/SECURITY.md)) for more information on security practices and recommendations associated with this project.

## Quick links

- Documentation: <https://github.com/software-gardening/almanac>
- Issue tracker: <https://github.com/software-gardening/almanac/issues>

## Process

### Reporting bugs or suggesting enhancements

We’re deeply committed to a smooth and intuitive user experience which helps people benefit from the content found within this project.
This commitment requires a good relationship and open communication with our users.

> We encourage you to report bugs or propose enhancements to improve the Software Gardening Almanac as a [GitHub issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-an-issue) associated with the repository.

Reviewer: ```suggestion
We encourage you to file a [GitHub issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-an-issue) to report bugs or propose enhancements to improve the Software Gardening Almanac.
```

Optional suggestion - this may be slightly clearer.

Author: Love it, thanks!

> First, figure out if your proposal is already implemented by reading existing issues or pull requests!

Reviewer: ```suggestion
First, figure out if your idea is already implemented by reading existing issues or pull requests!
```

> Next, check the issues (<https://github.com/software-gardening/almanac/issues>) to see if someone else has already reported the bug or proposed the enhancement you have in mind.

Reviewer: ```suggestion
Check the issues (<https://github.com/software-gardening/almanac/issues>) and pull requests (<https://github.com/software-gardening/almanac/pulls>)  to see if someone else has already documented or began implementation of your idea.
```

> If you do find the suggestion, please comment on the existing issue noting that you are also interested in the functionality.

Reviewer: ```suggestion
If you do find your idea in an existing issue, please comment on the existing issue noting that you are also interested in the functionality.
```

> If you do not find the suggestion, please open a new issue and document the specific enhancement and why it would be helpful for your particular use case.

Reviewer: ```suggestion
If you do not find your idea, please open a new issue and document why it would be helpful for your particular use case.
```

> Specifically, the following may help when it comes to addressing the issue:

Reviewer: ```suggestion
Please also provide the following specifics:
```

> - The version of the Software Gardening Almanac you're referencing.

Reviewer: It might be helpful to add this as an ISSUE_TEMPLATE as well, since many will not thoroughly read the CONTRIBUTING guide

Author: Great idea, thanks, I'll add these into an ISSUE_TEMPLATE with another addition!

We specifically follow a [forked pull request model](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork).
Please create a fork of Software Gardening Almanac repository, clone the fork, and then create a new, feature-specific branch.
Once you make the necessary changes on this branch, you should file a pull request to incorporate your changes into the main The Software Gardening Almanac repository.

The content and description of your pull request are directly related to the speed at which we are able to review, approve, and merge your contribution into The Software Gardening Almanac.
To ensure an efficient review process please perform the following steps:

1. Follow all instructions in the [pull request template](https://github.com/software-gardening/almanac/blob/main/.github/PULL_REQUEST_TEMPLATE.md)
1. Triple check that your pull request is adding _one_ specific feature. Small, bite-sized pull requests move so much faster than large pull requests.
1. After submitting your pull request, ensure that your contribution passes all status checks (e.g. passes all tests)

> Pull request review and approval is required by at least one project maintainer to merge.

Reviewer: perfect!

We will do our best to review the code additions in a timely fashion.
Ensuring that you follow all steps above will increase our speed and ability to review.
We will check for accuracy, style, code coverage, and scope.

### Git commit messages

For all commit messages, please use a short phrase that describes the specific change.
For example, “Add feature to check normalization method string” is much preferred to “change code”.
When appropriate, reference issues (via `#` plus number) .

## Development

### Overview

> The Software Gardening Almanac is primarily written in [Python](https://www.python.org/) through [Jupyter Book](https://jupyterbook.org/) with related environments managed by Python [Poetry](https://python-poetry.org/).
> [Node](https://nodejs.org) and [NPM](https://www.npmjs.com/) dependencies are also used to assist development activities.

Reviewer: ```suggestion
We write and develop the Software Gardening Almanac in [Python](https://www.python.org/) through [Jupyter Book](https://jupyterbook.org/) with related environments managed by Python [Poetry](https://python-poetry.org/).
We use [Node](https://nodejs.org) and [NPM](https://www.npmjs.com/) dependencies to assist development activities.
```

To move to active voice. Also, it might be worth clarifying (or providing an example) of the development activities.

Author: Thanks, good point! I'll add this information in with more content.

We use [GitHub actions](https://docs.github.com/en/actions) for [CI/CD](https://en.wikipedia.org/wiki/CI/CD) procedures such as automated tests.

### Getting started

To enable local development, perform the following steps.

1. [Install Python](https://www.python.org/downloads/) (suggested: use [`pyenv`](https://github.com/pyenv/pyenv) for managing Python versions)
1. [Install Poetry](https://python-poetry.org/docs/#installation)
1. [Install Poetry environment](https://python-poetry.org/docs/basic-usage/#installing-dependencies): `poetry install`
1. [Install Node](https://nodejs.org/en/download) (suggested: use [`nvm`](https://github.com/nvm-sh/nvm) for managing Node versions)
1. [Install Node environment](https://docs.npmjs.com/cli/v9/commands/npm-install): `npm install`

### Development Tasks

> Common development tasks are defined using [Poe the Poet](https://poethepoet.natn.io/), which helps simplify repeated use of commands.
> Poe the Poet is included as a Python Poetry dev group dependency and tasks may be used through the Poetry environment.

Reviewer: ```suggestion
We use [Poe the Poet](https://poethepoet.natn.io/) to define common development tasks, which simplifies repeated  commands.
We include Poe the Poet as a Python Poetry dev group dependency, which users access through the Poetry environment.
```

It might be helpful to provide an example of "repeated commands"

Reviewer: Ah! You do provide a link to the available tasks.

Reviewer: And another concrete example, great!

Author: Thanks! 🙂

Please see the [`pyproject.toml`](pyproject.toml) file's `[tool.poe.tasks]` table for a list of available tasks.

For example:

```shell
# example of how Poe the Poet commands may be used.
poetry run poe <task_name>

# example of a task which runs jupyter-book build for the project
poetry run poe build-book
```

### Linting

> Work added to this repo is automatically checked using [pre-commit](https://pre-commit.com/) via [GitHub Actions](https://docs.github.com/en/actions).

Reviewer: ```suggestion
[pre-commit](https://pre-commit.com/) automatically checks all work added to this repo.
We implement these checks using [GitHub Actions](https://docs.github.com/en/actions).
```

Pre-commit can work alongside your local [git with git-hooks](https://pre-commit.com/index.html#3-install-the-git-hook-scripts)

> After [installing pre-commit](https://pre-commit.com/#installation) within your development environment, the following command also can perform the same checks:

Reviewer: ```suggestion
After [installing pre-commit](https://pre-commit.com/#installation) within your development environment, the following command performs the same checks:
```

```sh

> % pre-commit run --all-files

Reviewer: Is this in the PULL_REQUEST template? If not, consider adding - I think it could be a great addition to the checklist!

Author: Great idea, thank you, I'll get this added!

```

### Automated CI/CD with GitHub Actions

We use [GitHub Actions](https://docs.github.com/en/actions) to help perform automated [CI/CD](https://en.wikipedia.org/wiki/CI/CD) as part of this project.
GitHub Actions involves defining [workflows](https://docs.github.com/en/actions/using-workflows) through [YAML files](https://en.wikipedia.org/wiki/YAML).

> These workflows include one or more [jobs](https://docs.github.com/en/actions/using-jobs) which are collections of steps, individual processes which run as part of a job.

Reviewer: ```suggestion
These workflows include one or more [jobs](https://docs.github.com/en/actions/using-jobs) which are collections of individual processes (or steps) which run as part of a job.
```

> GitHub Actions work is defined under the [`.github`](.github) directory.

Reviewer: ```suggestion
We define GitHub Actions work under the [`.github`](.github) directory.
```
